Dockerfile

```diff
@@ -1,42 +1,82 @@
-FROM golang:1.17-alpine as cloudreve_builder
+# 多阶段构建: 编译发布物
+# 注意: 该阶段除最终发布物外不会影响最终镜像体积, 删除缓存、减少包安装等无意义
+FROM golang:1-alpine AS builder
 
+# 安装基本编译依赖
+RUN set -ex \
+    && apk add build-base bash git yarn zip \
+    && git clone --recursive https://github.com/cloudreve/Cloudreve.git /source
 
-# install dependencies and build tools
-RUN apk update && apk add --no-cache wget curl git yarn build-base gcc abuild binutils binutils-doc gcc-doc zip
+# 编译前端代码
+WORKDIR /source/assets
 
-WORKDIR /cloudreve_builder
-RUN git clone --recurse-submodules https://github.com/cloudreve/Cloudreve.git
+# 允许通过 `docker build --build-args YARN_REGISTRY=https://registry.npmmirror.com` 覆盖默认仓库地址
+#
+# 常用的仓库地址:
+#   npm -----  https://registry.npmjs.org
+#   cnpm ----  http://r.cnpmjs.org
+#   taobao --  https://registry.npmmirror.com
+#   nj ------  https://registry.nodejitsu.com
+#   rednpm --  http://registry.mirror.cqupt.edu.cn
+#   skimdb --  https://skimdb.npmjs.com/registry
+#   yarn ----  https://registry.yarnpkg.com
+ARG YARN_REGISTRY=https://registry.yarnpkg.com
 
-# build frontend
-WORKDIR /cloudreve_builder/Cloudreve/assets
-RUN yarn install --network-timeout 1000000
-RUN yarn run build && find . -name "*.map" -type f -delete
+# 允许通过 `docker build --build-args GOPROXY=https://goproxy.cn` 添加 go mod 代理
+ARG GOPROXY=""
 
-# build backend
-WORKDIR /cloudreve_builder/Cloudreve
-RUN zip -r - assets/build >assets.zip
-RUN tag_name=$(git describe --tags) \
-    && export COMMIT_SHA=$(git rev-parse --short HEAD) \
-    && go build -a -o cloudreve -ldflags " -X 'github.com/HFO4/cloudreve/pkg/conf.BackendVersion=$tag_name' -X 'github.com/HFO4/cloudreve/pkg/conf.LastCommit=$COMMIT_SHA'"
+# 暂不确定未来 alpine 内的 Node 版本是否影响最终编译结果, 故暂时增加打印输出
+RUN set -ex \
+    && echo "Node Version: $(node -v)" \
+    && sed -i -e "s@https://registry.yarnpkg.com@${YARN_REGISTRY}@g" yarn.lock \
```

Comment:
@HFO4  I think we shouldn't do that replacement directly based on our former test?
@mritd  please make sure that works

Reply:
Yes, this is quite a hacky way.